src/Auth0Client.ts

```diff
@@ -493,4 +488,89 @@ export default class Auth0Client {
     const url = this._url(`/v2/logout?${createQueryParams(logoutOptions)}`);
     window.location.assign(`${url}${federatedQuery}`);
   }
+
+  private async _getTokenFromIFrame(
+    options: GetTokenSilentlyOptions
+  ): Promise<any> {
+    const stateIn = encodeState(createRandomString());
+    const nonceIn = createRandomString();
+    const code_verifier = createRandomString();
+    const code_challengeBuffer = await sha256(code_verifier);
+    const code_challenge = bufferToBase64UrlEncoded(code_challengeBuffer);
+
+    const authorizeOptions = {
+      audience: options.audience,
+      scope: options.scope
+    };
+
+    const params = this._getParams(
+      authorizeOptions,
+      stateIn,
+      nonceIn,
+      code_challenge,
+      this.options.redirect_uri || window.location.origin
+    );
+
+    const url = this._authorizeUrl({
+      ...params,
+      prompt: 'none',
+      response_mode: 'web_message'
+    });
+
+    const codeResult = await runIframe(url, this.domainUrl);
+
+    if (stateIn !== codeResult.state) {
+      throw new Error('Invalid state');
+    }
+
+    const tokenResult = await oauthToken({
+      baseUrl: this.domainUrl,
+      audience: options.audience || this.options.audience,
+      client_id: this.options.client_id,
+      code_verifier,
+      code: codeResult.code,
+      grant_type: 'authorization_code'
+    } as OAuthTokenOptions);
+
+    const decodedToken = this._verifyIdToken(tokenResult.id_token, nonceIn);
+
+    return {
+      ...tokenResult,
+      decodedToken,
+      scope: params.scope,
+      audience: params.audience || 'default'
+    };
+  }
+
+  private async _getTokenUsingRefreshToken(
+    options: GetTokenSilentlyOptions
+  ): Promise<any> {
+    const cache = this.cache.get({
+      scope: options.scope,
+      audience: options.audience || 'default'
+    });
+
+    if (!cache || !cache.refresh_token) {
+      throw new GenericError(
+        'missing_refresh_token',
+        'No refresh token is available to fetch a new access token. The user should be reauthenticated.'
+      );
+    }
+
+    const tokenResult = await oauthToken({
+      baseUrl: this.domainUrl,
+      client_id: this.options.client_id,
+      grant_type: 'refresh_token',
+      refresh_token: cache.refresh_token
+    } as RefreshTokenOptions);
```

Comment:
Why `OAuthTokenOptions` and `RefreshTokenOptions`? I know the former already existed but seems OK to deprecated or just stop using.

Reply:
These have actually been refactored into a hierarchy. So, `OAuthTokenOptions` did exist previously but had _all_ of the options on it. I've changed it so that both `OAuthTokenOptions` and `RefreshTokenOptions` both inherit from `TokenEndpointOptions`, which has all the common stuff on it.

Then, `OAuthTokenOptions` has all the stuff that's relevant for calling the token endpoint for the `authorization_code` grant (code verifier, etc) and `RefreshTokenOptions` has the `refresh_token` option available.

Make sense? I'm not married to the names but I think the hierarchy works.

Reply:
The hierarchy makes sense, I'm talking more about the names here. We're doing OAuth either way, getting tokens either way. Seems like `IframeTokenOptions` would make more sense to match the method.

Reply:
Not blocking :)

Reply:
I get you. I'd like to rename this to `PCKETokenOptions` or something more applicable to using the `authorization_code` grant (which is when this type is used). I'm going to do that in a separate PR though, I want to do some more investigation as to whether that would be a breaking change or not.